include/nbl/core/containers/DoublyLinkedList.h

```diff
@@ -0,0 +1,330 @@
+	inline const node_t* getBegin() const { return m_array + m_begin; }
+
+	//get node ptr of the last item in the list
+	inline node_t* getBack() { return m_array + m_back; }
+	inline const node_t* getBack() const { return m_array + m_back; }
+
+	//get index/iterator of the first element
+	inline uint32_t getFirstAddress() const { return m_begin; }
+
+	//get index/iterator of the last element
+	inline uint32_t getLastAddress() const { return m_back; }
+
+	//add new item to the list. This function does not make space to store the new node. in case the list is full, popBack() needs to be called beforehand
+	inline void pushFront(value_t&& val)
+	{
+		insertAt(reserveAddress(), std::move(val));
+	}
+
+	template <typename... Args>
+	inline void emplaceFront(Args&&... args)
+	{
+		insertAt(reserveAddress(), value_t(std::forward<Args>(args)...));
+	}
+
+	/**
+	* @brief Resets list to initial state.
+	*/
+	inline void clear()
+	{
+		disposeAll();
+
+		m_addressAllocator = std::unique_ptr<address_allocator_t>(new address_allocator_t(m_reservedSpace, 0u, 0u, 1u, m_cap, 1u));
+		m_back = invalid_iterator;
+		m_begin = invalid_iterator;
+
+	}
+
+	//remove the last element in the list
+	inline void popBack()
+	{
+		if (m_back == invalid_iterator)
+			return;
+
+		auto backNode = getBack();
+		if (backNode->prev != invalid_iterator)
+			get(backNode->prev)->next = invalid_iterator;
+		uint32_t temp = m_back;
+		m_back = backNode->prev;
+		common_delete(temp);
+	}
+
+	//remove a node at nodeAddr from the list
+	inline void erase(const uint32_t nodeAddr)
+	{
+		assert(nodeAddr != invalid_iterator);
+		assert(nodeAddr < m_cap);
+		node_t* node = get(nodeAddr);
+
+		if (m_back == nodeAddr)
+			m_back = node->prev;
+		if (m_begin == nodeAddr)
+			m_begin = node->next;
+
+		common_detach(node);
+		common_delete(nodeAddr);
+	}
+
+	//move a node at nodeAddr to the front of the list
+	inline void moveToFront(const uint32_t nodeAddr)
+	{
+		if (m_begin == nodeAddr || nodeAddr == invalid_iterator)
+			return;
+
+		getBegin()->prev = nodeAddr;
+
+		auto node = get(nodeAddr);
+
+		if (m_back == nodeAddr)
+			m_back = node->prev;
+
+		common_detach(node);
+		node->next = m_begin;
+		node->prev = invalid_iterator;
+		m_begin = nodeAddr;
+	}
+
+	/**
+	* @brief Resizes the list by extending its capacity so it can hold more elements. Returns a bool indicating if capacity was indeed increased.
+	*
+	* @param [in] newCapacity New number of elements to hold. MUST be greater than current list capacity.
+	*/
+	inline bool grow(uint32_t newCapacity)
+	{
+		// Must at least make list grow
+		if (newCapacity <= m_cap)
+			return false;
+		// Same as code found in ContiguousMemoryLinkedListBase to create aligned space
+		const auto firstPart = core::alignUp(address_allocator_t::reserved_size(1u, newCapacity, 1u), alignof(node_t));
+		void* newReservedSpace = _NBL_ALIGNED_MALLOC(firstPart + newCapacity * sizeof(node_t), alignof(node_t));
```

Comment:
like every good container take an `allocator` template param, and use it instead of malloc